# enforce-functions

Automatically transform lodash functions to their native JavaScript equivalents with optional include/exclude policies.

| | |
|:---|:---|
| **Rule type** | problem |
| **Fixable** | Yes |
| **Recommended** | No |

## Rule Details

This rule automatically transforms lodash function calls to their native JavaScript equivalents, providing modern, performant alternatives. It supports 67+ lodash functions with comprehensive autofix capabilities including ES2022+ features.

**Primary Features:**

- üîß **Automatic transformation** to native JavaScript
- üì¶ **Bundle size reduction** by eliminating lodash dependencies
- üöÄ **Modern JavaScript** support (ES2022 Array.at(), ES2024 Object.groupBy())
- ‚öôÔ∏è **Configurable policies** via include/exclude options

**Note**: You cannot use both `include` and `exclude` options together. Use one or the other.

## Automatic Transformations

The rule automatically transforms lodash functions to their native equivalents:

### Array Functions

```typescript
// Before (lodash)
import { map, filter, first, last } from 'lodash-es'

const doubled = map([1, 2, 3], x => x * 2)
const evens = filter([1, 2, 3, 4], x => x % 2 === 0)
const firstItem = first(items)
const lastItem = last(items)

// After (native - automatic transformation)
import { map, filter, first, last } from 'lodash-es'

const doubled = [1, 2, 3].map(x => x * 2)
const evens = [1, 2, 3, 4].filter(x => x % 2 === 0)
const firstItem = items.at(0)          // ES2022 Array.at()
const lastItem = items.at(-1)          // Negative indexing
```

### Modern JavaScript Features

```typescript
// Before (lodash)
import { groupBy, sortBy, cloneDeep } from 'lodash-es'

const grouped = groupBy(users, 'department')
const sorted = sortBy(data, 'score')
const deep = cloneDeep(complex)

// After (native - automatic transformation)
import { groupBy, sortBy, cloneDeep } from 'lodash-es'

const grouped = Object.groupBy(users, user => user.department)  // ES2024
const sorted = data.toSorted((a, b) => a.score - b.score)      // ES2023
const deep = structuredClone(complex)                          // Native API
```

### Complex Transformations

```typescript
// Before (lodash)
import { has, pick, reject } from 'lodash-es'

const hasName = has(user, 'name')
const subset = pick(user, ['name', 'email'])
const nonAdmins = reject(users, user => user.role === 'admin')

// After (native - automatic transformation)
import { has, pick, reject } from 'lodash-es'

const hasName = 'name' in user
const subset = Object.fromEntries(['name', 'email'].map(k => [k, user[k]]))
const nonAdmins = users.filter(item => !(user => user.role === 'admin')(item))
```

## Configuration Options

When functions are excluded or not included, the rule flags them without transformation:

```typescript
// Configuration: { "exclude": ["forEach", "map"] }
import { forEach, map, filter } from 'lodash-es'

forEach(items, callback) // ‚ùå forEach is excluded - flagged, no transformation
map(items, mapper)       // ‚ùå map is excluded - flagged, no transformation
filter(items, predicate) // ‚úÖ filter transforms to: items.filter(predicate)
```

## Supported Functions (67+)

### Array Functions (19 functions)

- `map` ‚Üí `array.map(fn)`
- `filter` ‚Üí `array.filter(predicate)`
- `find` ‚Üí `array.find(predicate)`
- `reduce` ‚Üí `array.reduce(fn, initial)`
- `forEach` ‚Üí `array.forEach(fn)`
- `some` ‚Üí `array.some(predicate)`
- `every` ‚Üí `array.every(predicate)`
- `slice` ‚Üí `array.slice(start, end)`
- `concat` ‚Üí `array.concat(...values)`
- `join` ‚Üí `array.join(separator)`
- `indexOf` ‚Üí `array.indexOf(value)`
- `lastIndexOf` ‚Üí `array.lastIndexOf(value)`
- `flatten` ‚Üí `array.flat()`
- `flatMap` ‚Üí `array.flatMap(fn)`
- `reduceRight` ‚Üí `array.reduceRight(fn, initial)`
- `first` ‚Üí `array.at(0)` (ES2022)
- `last` ‚Üí `array.at(-1)` (ES2022)
- `includes` ‚Üí `array.includes(value)`
- `reverse` ‚Üí `array.reverse()` ‚ö†Ô∏è mutates

### Object Functions (4 functions)

- `keys` ‚Üí `Object.keys(object)`
- `values` ‚Üí `Object.values(object)`
- `entries` ‚Üí `Object.entries(object)`
- `assign` ‚Üí `Object.assign(target, ...sources)`

### String Functions (12 functions)

- `trim` ‚Üí `string.trim()`
- `trimStart` ‚Üí `string.trimStart()`
- `trimEnd` ‚Üí `string.trimEnd()`
- `toLower` ‚Üí `string.toLowerCase()`
- `toUpper` ‚Üí `string.toUpperCase()`
- `startsWith` ‚Üí `string.startsWith(target)`
- `endsWith` ‚Üí `string.endsWith(target)`
- `repeat` ‚Üí `string.repeat(n)`
- `replace` ‚Üí `string.replace(pattern, replacement)`
- `split` ‚Üí `string.split(separator)`
- `padStart` ‚Üí `string.padStart(length, chars)`
- `padEnd` ‚Üí `string.padEnd(length, chars)`

### Collection Functions (11 functions)

- `reject` ‚Üí `array.filter(item => !predicate(item))`
- `size` ‚Üí `value.length`
- `groupBy` ‚Üí `Object.groupBy(array, fn)` (ES2024)
- `countBy` ‚Üí reduce pattern for counting
- `keyBy` ‚Üí `Object.fromEntries(array.map(...))`
- `chunk` ‚Üí `Array.from()` pattern
- `partition` ‚Üí dual filter operations
- `orderBy` ‚Üí `array.toSorted(...)` (ES2023)
- `each` ‚Üí `array.forEach(fn)`
- `findLast` ‚Üí `array.findLast(predicate)` (ES2023)
- Complex object manipulation functions

### Type Checking Functions (11 functions)

- `isString` ‚Üí `typeof value === "string"`
- `isNumber` ‚Üí `typeof value === "number"`
- `isBoolean` ‚Üí `typeof value === "boolean"`
- `isFunction` ‚Üí `typeof value === "function"`
- `isObject` ‚Üí `typeof value === "object" && value !== null`
- `isNull` ‚Üí `value === null`
- `isUndefined` ‚Üí `value === undefined`
- `isNil` ‚Üí `value == null`
- `isArray` ‚Üí `Array.isArray(value)`
- `toString` ‚Üí `value.toString()`
- `toNumber` ‚Üí `Number(value)`

### Math/Number Functions (8 functions)

- `isFinite` ‚Üí `Number.isFinite(value)`
- `isInteger` ‚Üí `Number.isInteger(value)`
- `isNaN` ‚Üí `Number.isNaN(value)`
- `max` ‚Üí `Math.max(...array)`
- `min` ‚Üí `Math.min(...array)`
- `ceil` ‚Üí `Math.ceil(number)`
- `floor` ‚Üí `Math.floor(number)`
- `round` ‚Üí `Math.round(number)`

### Advanced Functions

- `has` ‚Üí `key in object` (parameter reordering)
- `pick` ‚Üí `Object.fromEntries()` pattern
- `omit` ‚Üí `Object.entries().filter()` pattern
- `merge` ‚Üí `Object.assign({}, ...)` pattern
- `get` ‚Üí optional chaining `obj?.prop?.path`
- `clone` ‚Üí `{...obj}` spread
- `cloneDeep` ‚Üí `structuredClone(obj)` (native browser API)
- `uniq` ‚Üí `[...new Set(array)]`
- `compact` ‚Üí `array.filter(Boolean)`
- `sortBy` ‚Üí `array.toSorted(...)` (ES2023)
- `now` ‚Üí `Date.now()`

## Options

This rule accepts an options object with the following properties:

- `include` (array of strings): List of lodash functions to allow. If provided, only these functions are permitted.
- `exclude` (array of strings): List of lodash functions to disallow/block.

### exclude

Examples of **incorrect** code for the `exclude` option:

```typescript
/* eslint lodash-es/enforce-functions: ["error", { "exclude": ["forEach", "map"] }] */
import { forEach, map } from 'lodash-es'

forEach(items, callback) // ‚ùå forEach is excluded
map(items, mapper)       // ‚ùå map is excluded
```

Examples of **correct** code for the `exclude` option:

```typescript
/* eslint lodash-es/enforce-functions: ["error", { "exclude": ["forEach"] }] */
import { map, filter } from 'lodash-es'

map(items, mapper)       // ‚úÖ map is not excluded
filter(items, predicate) // ‚úÖ filter is not excluded
```

### include

Examples of **incorrect** code for the `include` option:

```typescript
/* eslint lodash-es/enforce-functions: ["error", { "include": ["map", "filter"] }] */
import { map, filter, reduce } from 'lodash-es'

reduce(items, reducer)   // ‚ùå reduce is not included
```

Examples of **correct** code for the `include` option:

```typescript
/* eslint lodash-es/enforce-functions: ["error", { "include": ["map", "filter"] }] */
import { map, filter } from 'lodash-es'

map(items, mapper)       // ‚úÖ map is included
filter(items, predicate) // ‚úÖ filter is included
```

## Use Cases

### Performance-focused teams

Exclude functions with performance implications:

```json
{
  "lodash-es/enforce-functions": ["error", {
    "exclude": ["forEach", "forEachRight", "forIn", "forInRight"]
  }]
}
```

### Functional programming style

Include only functional programming utilities:

```json
{
  "lodash-es/enforce-functions": ["error", {
    "include": [
      "map", "filter", "reduce", "find", "some", "every",
      "flatten", "flatMap", "uniq", "groupBy", "keyBy"
    ]
  }]
}
```

### Gradual migration

Exclude functions that have native alternatives:

```json
{
  "lodash-es/enforce-functions": ["error", {
    "exclude": ["forEach", "map", "filter", "find", "some", "every"]
  }]
}
```

## Enhanced Error Messages

This rule provides enhanced error messages that include native alternatives when available:

```
Lodash function 'forEach' is excluded by configuration. Consider using native Array.prototype.forEach: array.forEach(callback)
```

## When Not To Use It

You might want to disable this rule if:

- You want to allow all lodash functions without restrictions
- You're in early development and haven't decided on function policies yet
- You're migrating from another lodash plugin and need flexibility

## Related Rules

- [enforce-destructuring](./enforce-destructuring.md) - Enforce destructured imports
- [suggest-native-alternatives](./suggest-native-alternatives.md) - Suggests native JavaScript alternatives

## Version

This rule was introduced in eslint-plugin-lodash-es v0.1.0.

## Resources

- [Rule source](../../src/rules/enforce-functions.ts)
- [Test source](../../tests/enforce-functions.test.ts)
